File: July-LeetCoding-Challenge-2020/Week1/441-Arranging-Coins.cpp
// Solution #1 Brute Force
// TC: O(n)
// SC: O(1)
// Runtime: 8 ms, faster than 42.54%
// Memory Usage: 6.1 MB, less than 10.30%
class Solution {
public:
    
    int arrangeCoins(int n) {
        
        int res = 0;
        int coins = 1;
        
        while(1){
            
            n -= coins++;
            
            if(n >= 0){
               res++;
            }else{
                return res;
            }
        }
        
        return -1;
    }
};

// Solution #2 Math
// TC: O(1)
// SC: O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 5.9 MB, less than 56.91%
class Solution {
public:
    
    int arrangeCoins(int n) {
        
        return (-1 + sqrt(1 + (long)n*8))/2;
    }
};

// Solution #3 Binary Search
// TC: O(logn)
// SC: O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 5.9 MB, less than 56.91%
class Solution {
public:
    
    int arrangeCoins(int n) {
        
        long left = 1, right = n;
        long K;
        long coins;
        
        while(left <= right){
            K = (left + right)/2;
            coins = (K*(K+1))/2;
            
            if(coins > n){
                right = K - 1;
            }else{
                left = K + 1;
            }
        }
        
        return left-1;
    }
};

File: July-LeetCoding-Challenge-2020/Week1/441-arranging-coins.py
class Solution:
    # Sol 1. One Pointer
    def arrangeCoins(self, n: int) -> int:
        stair_count = 1
        while stair_count <= n:
            n, stair_count = n-stair_count, stair_count+1
            # print(stair_count)
        return stair_count-1
    # TC: O(n)
    # SC: O(1)
    # Runtime: 904 ms, faster than 40.41%
    # Memory Usage: 13.6 MB, less than 96.42%
    
    # Sol 2. Math
    def arrangeCoins(self, n: int) -> int:
        return abs(math.floor(math.sqrt(2*n+0.25) - 0.5))
    # TC: O(1)
    # SC: O(1)
    # Runtime: 32 ms, faster than 86.95%
    # Memory Usage: 14 MB, less than 14.92%

    # Sol 3. Binary Search
    def arrangeCoins(self, n: int) -> int:
        # # coins = k*(k+1)//2
        l, r = 0, n
        
        while l<=r:
            mid = (l+r)//2
            coins = ((mid)*(mid+1))//2
            if coins == n:
                return mid
            elif coins > n:
                r = mid - 1
            else:
                l = mid + 1
        return r # maxmize the # stairs
    # TC: O(logn)
    # SC: O(1)
    # Runtime: 36 ms, faster than 75.73%
    # Memory Usage: 13.6 MB, less than 98.64%
